### DISCUSSION TITLE
How do you reload & sync commands in cogs?

### DISCUSSION THREAD

**Discussion by ItsCubeTime** [Rapptz/discord.py#9746] [Q&A]
I have an extension `modpinmsgtobottom`, with a cog `ModPinMsgToBottomPy_COG`:
```py
import discord
from discord.ext import commands

async def setup(client: commands.Bot):
    await client.add_cog(ModPinMsgToBottomPy_COG(client),override=True)

class ModPinMsgToBottomPy_COG(commands.Cog):
    def __init__(self, client: discord.Client):
        # client.
        print(f"Initializing cog '{self.__cog_name__}'")
        self.client = client

    @commands.Cog.listener()
    async def on_member_join(self, member):
        channel = member.guild.system_channel
        if channel is not None:
            await channel.send(f'Welcome {member.mention}.')
      


    @discord.app_commands.command()
    async def hello8(self, interaction: discord.Interaction, text: str):
        """Says hello"""
        await interaction.response.send_message("meow8")
```
And another extension `developmentUtilitiesCog` with a cog `DevelopmentUtilities_Cog` which defines a command `reloadextension `:
```py
import discord
from discord.ext import commands

async def setup(client: commands.Bot):
    await client.add_cog(DevelopmentUtilities_Cog(client))

class DevelopmentUtilities_Cog(commands.Cog):
    def __init__(self, client: commands.Bot):
        print(f"Initializing cog '{self.__cog_name__}'")
        self.client = client
    @discord.app_commands.command()
    async def reloadextension(self,interaction: discord.Interaction,cogname:str=""):
        """Says hello"""
        if cogname=="":
            keys = list(self.client.extensions.keys())
            for extension in keys:
                await self.client.reload_extension(extension)
                await self.client.unload_extension(extension)
                await self.client.load_extension(extension)
        else:
            await self.client.reload_extension(cogname)
            await self.client.unload_extension(cogname)
            await self.client.load_extension(cogname)
        for cmd in self.get_commands():
            if not cmd.name == 'reloadextension':
                cmd.update()
        # for _guild in self.client.guilds:
            # print(f"Attempting to sync commands with server {_guild.name}")
        await self.client.tree.sync()
        for _guild in self.client.guilds:
            print(f"Attempting to sync commands with server {_guild.name}")
            await self.client.tree.sync(guild = _guild)
            print(f"Success!")
        await interaction.response.send_message(f"Reloading cog {cogname}" if cogname!="" else "Reloading all cogs")
```

Running `/reloadextension modpinmsgtobottom` is supposed to hot reload the code in my `modpinmsgtobottom`  extension (which it seems to do just fine, in that I can change function definitions and changes take effect immediately).

The only issue is that the commands doesnt seem to sync (I cant see new commands without restarting the entire bot). Am I missing something?

### Thanks for reading 💝

**Comment by ItsCubeTime** [comment]
Bump, I don't think I got this working

**Comment by lexicalunit** [comment]
To get things work for me I need to do something like this:
```python
from importlib import import_module

bot: AutoShardedBot

bot.tree.clear_commands()
await bot.tree.sync()

# prevent rate limiting:
await asyncio.sleep(1)

# your cog modules:
cogs = ["cogs.foo", "cogs.bar", "cogs.baz"]

# load or reload each of them:
for cog in cogs:
    module = import_module(cog)
    if module.__name__ in bot.extensions:
        await bot.reload_extension(module.__name__)
    else:
        await bot.load_extension(module.__name__)

await bot.tree.sync()
```

**Reply by ItsSNOWMOBILE** [reply]
This is a bit old but why are you manually loading cogs?
```py
for filename in os.listdir("./cogs"):
        if filename.endswith(".py"):
            cog_name = f"cogs.{filename[:-3]}"
            try:
                bot.load_extension(cog_name)
```

**Reply by lexicalunit** [reply]
I found that in order for the sync to work, you need to load/reload the cogs. Without this code my bot doesn't have any commands when I run it.

**Comment by AbstractUmbra** [comment]
For Discord to be updated with changes to your application commands you must [sync](https://discordpy.readthedocs.io/en/stable/interactions/api.html#discord.app_commands.CommandTree.sync) the [CommandTree](https://discordpy.readthedocs.io/en/stable/interactions/api.html#discord.app_commands.CommandTree) to inform them of these changes.
Upon syncing commands, it will invalidate your existing ones in cases of there being an update to them (e.g. name/description/params/names etc). This means you'll often need to reload your discord client with CTRL + R or quit and open it again.

I digress - the reason your commands don't sync when you reload an extension is because discord.py has no means or toggle for automatically syncing your CommandTree, this is entirely on you - the owner/developer - to do on demand as needed.

I recommend you join the discord server and open a help thread to discuss your means of finding and loading extensions, it isn't broken nor bad but does leave things to be desired for usability in developer experience, but again - not related to the current issue.
